src/util/simplify_expr2.cpp

```diff
@@ -1221,49 +1220,45 @@ static expr2tc do_bit_munge_operation(
   const expr2tc &side_2)
 {
   // Try to recursively simplify nested operations both sides, if any
-  expr2tc simplied_side_1 = try_simplification(side_1);
-  expr2tc simplied_side_2 = try_simplification(side_2);
+  expr2tc simplified_side_1 = try_simplification(side_1);
+  expr2tc simplified_side_2 = try_simplification(side_2);
 
-  if(!is_constant_expr(simplied_side_1) && !is_constant_expr(simplied_side_2))
+  /* Only support constant folding for integer and's. If you're a float,
+   * pointer, or whatever, you're on your own. */
+  if(
+    is_constant_int2t(simplified_side_1) &&
+    is_constant_int2t(simplified_side_2) && type->get_width() <= 64)
   {
-    // Were we able to simplify the sides?
-    if((side_1 != simplied_side_1) || (side_2 != simplied_side_2))
-    {
-      expr2tc new_op =
-        expr2tc(new constructor(type, simplied_side_1, simplied_side_2));
+    // So - we can't make BigInt by itself do the operation. But we can map it
+    // to the corresponding operation on our native types.
+    const constant_int2t &int1 = to_constant_int2t(simplified_side_1);
+    const constant_int2t &int2 = to_constant_int2t(simplified_side_2);
 
-      return typecast_check_return(type, new_op);
-    }
+    // Dump will zero-prefix and right align the output number.
+    int64_t val1 = int1.value.to_int64();
+    int64_t val2 = int2.value.to_int64();
 
-    return expr2tc();
-  }
-
-  // Only support integer and's. If you're a float, pointer, or whatever, you're
-  // on your own.
-  if(!is_constant_int2t(side_1) || !is_constant_int2t(side_2))
-    return expr2tc();
-
-  // So - we can't make BigInt by itself do an and operation. But we can dump
-  // it to a binary representation, and then and that.
-  const constant_int2t &int1 = to_constant_int2t(side_1);
-  const constant_int2t &int2 = to_constant_int2t(side_2);
+    uint64_t r = opfunc(val1, val2);
 
-  // Dump will zero-prefix and right align the output number.
-  int64_t val1 = int1.value.to_int64();
-  int64_t val2 = int2.value.to_int64();
+    if(type->get_width() < 64)
+    {
+      // truncate the result to the type's width
+      uint64_t trunc_mask = ~(uint64_t)0 << type->get_width();
+      r &= ~trunc_mask;
+      // if the type is signed and r's sign-bit is set, sign-extend it
+      if(is_signedbv_type(type) && r >> (type->get_width() - 1))
+        r |= trunc_mask;
+    }
+    return constant_int2tc(type, BigInt((int64_t)r));
+  }
```

Comment:
is this part of Lucas' PR? Maybe you need to rebase this branch?

Reply:
The diff is confusing, it's actually a different change: https://github.com/esbmc/esbmc/pull/581/commits/df7913aa29c73b6bb24b54f65c1189e319bfa2d1
I've moved the constant-folding up, so it will now a) do constant folding when the *simplified exprs* allow it (in contrast to the original ones) and otherwise b) not disregard a simplification that when one side is a constant_int2t.